### PR TITLE
feat: get variation details by variation type

### PR DESCRIPTION
**Pull request by duyhungtnn**
Related https://github.com/bucketeer-io/bucketeer/issues/886
### Changes
- Added a new data class `BKTEvaluationDetail<T>` to support getting variation details by variation type.
- Added a new data class `BKTValue` to reference [a JSON node value ](https://github.com/bucketeer-io/ios-client-sdk/pull/90/files#diff-a3a4eddf801ddca832aa6a192ebfb577b04567c5e6e08efa273d48be854ea950R6)
- Introduce a new interface to get variation details by variation type
```swift
func objectVariation(featureId: String, defaultValue: BKTValue) -> BKTValue
func intVariationDetails(featureId: String, defaultValue: Int) -> BKTEvaluationDetail<Int> {}
func doubleVariationDetails(featureId: String, defaultValue: Double) -> BKTEvaluationDetail<Double> {}
func boolVariationDetails(featureId: String, defaultValue: Bool) -> BKTEvaluationDetail<Bool> {}
func stringVariationDetails(featureId: String, defaultValue: String) -> BKTEvaluationDetail<String> {}
func objectVariationDetails(featureId: String, defaultValue: BKTValue) -> BKTEvaluationDetail<BKTValue> {}
```
- The class `BKTEvaluation` & the methods `getEvaluationDetails`, `jsonVariation` will be deprecated after this PR.

### CODE REVIEW

**Comment by duyhungtnn** [comment]
@cre8ivejp please help me to take a look on this PR

File: Bucketeer/Sources/Public/BKTClient.swift [hunk 1]
```diff
@@ -12,23 +12,34 @@ public class BKTClient {
         self.component = ComponentImpl(dataModule: dataModule)
     }
 
-    func getVariationValue<T>(featureId: String, defaultValue: T) -> T {
+    func getVariationValue<T: Equatable>(featureId: String, defaultValue: T) -> T {
+        return getVariationDetail(featureId: featureId, defaultValue: defaultValue).variationValue
+    }
+
+    func getVariationDetail<T:Equatable>(featureId: String, defaultValue: T) -> BKTEvaluationDetails<T> {
         component.config.logger?.debug(message: "BKTClient.getVariation(featureId = \(featureId), defaultValue = \(defaultValue) called")
```
Comment (kakcy) [inline Bucketeer/Sources/Public/BKTClient.swift]:
```suggestion
        component.config.logger?.debug(message: "BKTClient.getVariationDetail(featureId = \(featureId), defaultValue = \(defaultValue) called")
```

File: Bucketeer/Sources/Public/BKTEvaluation.swift [hunk 2]
```diff
@@ -19,3 +21,64 @@ public struct BKTEvaluation: Equatable {
...
+public struct BKTEvaluationDetails<T:Equatable>: Equatable {
+    public let featureId: String
+    public let featureVersion: Int
+    public let userId: String
+    public let variationId: String
+    public let variationName: String
+    public let variationValue: T
+    public let reason: Reason
+
+    public enum Reason: String, Codable, Hashable {
```
Comment (kakcy) [inline Bucketeer/Sources/Public/BKTEvaluation.swift]: Although it is similar to the [ReasonType](https://github.com/bucketeer-io/ios-client-sdk/blob/main/Bucketeer/Sources/Internal/Model/ReasonType.swift#L3C6-L3C16) definition, it is not very desirable to have the same definition in multiple places, so I would like you to consider using it uniformly if possible. Or consider removing the definition if you don't need it.

File: Bucketeer/Sources/Public/BKTEvaluation.swift [hunk 3]
```diff
@@ -19,3 +21,64 @@ public struct BKTEvaluation: Equatable {
...
+        case target = "TARGET"
+        case rule = "RULE"
+        case `default` = "DEFAULT"
+        case client = "CLIENT"
+        case offVariation = "OFF_VARIATION"
+        case prerequisite = "PREREQUISITE"
+
+        public static func fromString(value: String) -> Reason {
```
Comment (kakcy) [inline Bucketeer/Sources/Public/BKTEvaluation.swift]: It might be good to have a test for the `fromString()`.

File: Bucketeer/Sources/Public/BKTEvaluation.swift [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import Foundation
 
+@available(*, deprecated, message: "Use BKTEvaluationDetail<T> instead")
```
Comment (kakcy) [inline Bucketeer/Sources/Public/BKTEvaluation.swift]:
```suggestion
@available(*, deprecated, message: "Use BKTEvaluationDetails<T> instead")
```

File: Bucketeer/Sources/Public/BKTEvaluation.swift [hunk 5]
```diff
@@ -10,6 +11,7 @@ public struct BKTEvaluation: Equatable {
     public let variationValue: String
     public let reason: Reason
 
+    @available(*, deprecated, message: "Use BKTEvaluationDetail<T>.Reason instead")
```
Comment (kakcy) [inline Bucketeer/Sources/Public/BKTEvaluation.swift]:
```suggestion
    @available(*, deprecated, message: "Use BKTEvaluationDetails<T>.Reason instead")
```

File: Bucketeer/Sources/Public/BKTClient.swift [hunk 6]
```diff
@@ -200,6 +238,7 @@ extension BKTClient {
         }
     }
 
+    @available(*, deprecated, message: "use stringEvaluationDetails<T> instead")
```
Comment (kakcy) [inline Bucketeer/Sources/Public/BKTClient.swift]: ??
```suggestion
    @available(*, deprecated, message: "use getVariationDetail<T> instead")
```

**Review by kakcy — COMMENTED** [review]
thank you!
I have made some comments, please check them out.

File: Bucketeer/Sources/Public/BKTClient.swift (hunk 6 above)
Comment (duyhungtnn) [reply Bucketeer/Sources/Public/BKTClient.swift]: hi @kakcy , 

`getVariationDetail` is for internal use only. It can not be accessed from outside.
But I will remote the `<T>` from the `stringEvaluationDetails<T>`

File: Bucketeer/Sources/Public/BKTEvaluation.swift (hunk 2 above)
Comment (duyhungtnn) [reply Bucketeer/Sources/Public/BKTEvaluation.swift]: [ReasonType](https://github.com/bucketeer-io/ios-client-sdk/blob/main/Bucketeer/Sources/Internal/Model/ReasonType.swift#L3C6-L3C16)

is for internal use only. I maintain the same approach like before

File: Bucketeer/Sources/Public/BKTEvaluation.swift (hunk 3 above)
Comment (duyhungtnn) [reply Bucketeer/Sources/Public/BKTEvaluation.swift]: I will add them

Comment (duyhungtnn) [reply Bucketeer/Sources/Public/BKTEvaluation.swift]: updated in https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/d170ebe84a91bcc0b47b31cbfe310074d5e7cadd

File: Bucketeer/Sources/Public/BKTEvaluation.swift (hunk 2 above)
Comment (duyhungtnn) [reply Bucketeer/Sources/Public/BKTEvaluation.swift]: I would like to keep this. The Android SDK has something like this as well. @kakcy

File: Bucketeer/Sources/Public/BKTClient.swift (hunk 6 above)
Comment (duyhungtnn) [reply Bucketeer/Sources/Public/BKTClient.swift]: I updated it to 
```swift
@available(*, deprecated, message: "use stringVariationDetails(featureId:, defaultValue:) instead")
```

**Comment by duyhungtnn** [comment]
@kakcy thank you for reviewing.
I updated the PR with some changes from your suggestion.
Also, I changed the interface a little in this commit https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/12cbd20de5e303e36e8c4fce890ea8cbaf5fce39

as they are a new requirement https://github.com/bucketeer-io/android-client-sdk/pull/180#issuecomment-2188632955

**Review by kakcy — DISMISSED** [review]
Thank you for your responses 👍

**Review by cre8ivejp — DISMISSED** [review]
Thank you!
I'll merge next week.

**Comment by duyhungtnn** [comment]
hi @cre8ivejp , @kakcy 
Please help me to take a look at this PR. 
I added changes from the latest review. These changes make our interface closest to OpenFeatue specification 

# Changes 
- Added a new data class `BKTValue` to reference [a JSON node value](https://github.com/bucketeer-io/ios-client-sdk/pull/90/files#diff-a3a4eddf801ddca832aa6a192ebfb577b04567c5e6e08efa273d48be854ea950R6)
- Added [a new method](https://github.com/bucketeer-io/ios-client-sdk/pull/90/files#diff-2edc41fe19545933e618fc1376f5028e229b74f3e7c301ab02251976e3f915d3R162) to return a `BKTValue`
```
public func objectVariationDetails(featureId: String, defaultValue:BKTValue)
```
- Removed `jsonVariationDetail`
- Deprecated the method `jsonVariation`

**Comment by duyhungtnn** [comment]
Although this PR seems ready for review. However, I found some edge cases when finalizing the same changes in the Android PR.

I will add some new tests to this PR.

**Comment by duyhungtnn** [comment]
### New changes 
- Added support to return intValue even when the underlying value is a double https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/8d576f2c755d14578a826afcea8d8af9d42f3fd4
- Created BKTValue.number to hold JSON number value https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/58a6bb7a99e5792fcc17e2978b90ba4ac8d77baa

**Comment by duyhungtnn** [comment]
### New changes
- keep the detail interface right after the non-detail interface https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/3515534d3a7faae2f048afce54bad82c31c9dc76

File: Bucketeer/Sources/Public/BKTClient.swift [hunk 7]
```diff
@@ -126,24 +143,50 @@ extension BKTClient {
...
+        return getVariationDetail(featureId: featureId, defaultValue: defaultValue)
+    }
+
+    public func stringVariation(featureId: String, defaultValue: String) -> String {
+        return stringVariationDetails(featureId: featureId, defaultValue: defaultValue).variationValue
+    }
+
+    public func stringVariationDetails(featureId: String, defaultValue: String) -> BKTEvaluationDetails<String> {
+        return getVariationDetail(featureId: featureId, defaultValue: defaultValue)
+    }
```
Comment (cre8ivejp) [inline Bucketeer/Sources/Public/BKTClient.swift]: Can we call the same function for both cases?
It's a bit confusing. The non-details functions call the `stringVariationDetails` but the details function calls a different function `getVariationDetail`.

E.g.
```suggestion
    public func stringVariation(featureId: String, defaultValue: String) -> String {
        return stringVariationDetails(featureId: featureId, defaultValue: defaultValue).variationValue
    }

    public func stringVariationDetails(featureId: String, defaultValue: String) -> BKTEvaluationDetails<String> {
        return stringVariationDetails(featureId: featureId, defaultValue: defaultValue)
    }
```

Comment (cre8ivejp) [reply Bucketeer/Sources/Public/BKTClient.swift]: Like this.
https://github.com/bucketeer-io/go-server-sdk/pull/136/files#diff-9658208f158246a9cfbc2c798e77e8224eaed053c42d140eb1e7237d8ed4e91cR295-R305

File: Bucketeer/Sources/Public/BKTEvaluation.swift (hunk 2 above)
Comment (cre8ivejp) [inline Bucketeer/Sources/Public/BKTEvaluation.swift]: @duyhungtnn, WDYT of adding the `BKTEvaluationDetails` to a new file like you did in the Android SDK?

Comment (duyhungtnn) [reply Bucketeer/Sources/Public/BKTEvaluation.swift]: It's good.
I updated https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/6e761aca5f0edf4453bcf9e5778f0d7d418916b8

**Review by cre8ivejp — APPROVED** [review]
Thank you!
